### DISCUSSION TITLE
Add a simple way to keep aspect ratio of children of a GridContainer

### DISCUSSION THREAD

**Discussion by milana-94888u** [godotengine/godot-proposals#7198] [GUI]
Many games have something like level selection menu with multiple buttons in grid. They usually square. But in godot there's no way to compose Control nodes so they can do this. The problem with adding a AspectRatioContainer as a child node of a GridContainer is that AspectRatioContainer doesn't change own size but resize its children. So if I want to expand children of GridContainer to fill its width and I want to keep them square there's no simple way to properly align them vertically because of child nodes of AspectRatioContainer become bigger than a AspectRatioContainer itself and start cover each other added to a GridContainer. So that I want to propose to add something like AspectRatioGridContainer that will allow to keep the aspect of its children fixed and set their vertical sizing automatically

**Comment by juhaelee** [comment]
@milana-94888u did you ever find a solution or workaround to achieve this?

**Reply by Kartopod** [reply]
I'm not sure if I understand the problem correctly, but is this what you're trying to achieve?
![GridContainer](https://github.com/user-attachments/assets/0e716abe-2f12-41aa-b57a-46d3894d4de6)

Each of the buttons are nested under their own AspectRatioContainer, and all the AspectRatioContainers are set to FILL and EXPAND on both axes:
![image](https://github.com/user-attachments/assets/5eca15d9-ed32-435d-a969-5e0f15ba3f42)

**Reply by StockerGaming** [reply]
Is there a way to avoid the spaced between the items? I would much prefer them to be flush and to be aligned with the beginning, end or centre

**Reply by Calinou** [reply]
Spacing between items is defined in the GridContainer's theme items (H Separation and V Separation). Set them to 0 in the theme overrides:

<img width="391" height="115" alt="image" src="https://github.com/user-attachments/assets/d31cebaa-47ea-465d-b811-d202a487eeda" />

**Reply by StockerGaming** [reply]
Thank you, but shouldn’t the gaps that appear dynamically in the GIF still happen with both of those set to 0? As far as I am aware the extra space that is left is inside the AspectRatioContainers

**Comment by scraprats** [comment]
You want this right? A stacking uniform grid? It is possible using base controls (without code) but unfortunately it is a very inelegant solution.

![Screenshot_20241020_120227](https://github.com/user-attachments/assets/c9bf0304-3846-4ae6-ad1e-f7a9b539b7ad)

**Comment by StockerGaming** [comment]
I feel like something like this could be achieved by giving an option to AspectRatioContainers to fill with their respective ratio instead of seemingly not filling at all on their own

**Comment by StockerGaming** [comment]
Maybe this post should be more generalised to apply to (almost) all containers?
